field.go

```diff
@@ -410,6 +410,14 @@ func Inline(val zapcore.ObjectMarshaler) Field {
 	}
 }
 
+type anyFieldC[T any] func(string, T) Field
```

Comment:
can you add a comment about why we're doing all of this

ideally we'd have an upstream issue linked which describes the compiler issue

Reply:
Added more details. I don't think there's an upstream compiler issue yet. I believe there are plans to write one.

Reply:
Issue now exists, adding a link in https://github.com/uber-go/zap/pull/1387